keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -891,38 +891,68 @@ xref:
 - CRDs
 - Aggregated API servers
 
 Problem examples:
 
-| Problem                                                | Summary                                     |
-| ------------------------------------------------------ | ------------------------------------------- |
-| version skew: ephemeralContainers case                 | New pod field, need to be able to validate in same was containers and initContainers if field exists and is populated |
-| version skew: Migration from annotation to field       | Need to be able to validate annotation (if present) or field (if it exists and is populated) |
-| CRD is deleted                                         | Nothing to type check against, but also means there are no coresponding custom resources |
-| CRD is in multiple clusters, but schema differs        | If policy author is aware of the schema variations, can they write policies that work for all the variations? |
-| Validation of an aggregated API server type            | Main API server does not have type definitions |
-
-Due to these complications, we have decided to evaluate CEL expressions
-dynamically. Informational type checking will be provided (except for aggregated
-API server types), but will be surfaced only as warnings. See "Alternatives
-Considered" section for details of all the alternatives we reviewed when
-selecting this approach.
-
-Type checking is still performed for all expressions where a GVK can be matched
-to type check against, resulting in warnings, e.g.:
-
-```yaml
-...
-status:
-  expressionWarnings:
-    - expression: "object.foo"
-      warning: "no such field 'foo'"
-```
+| Problem                                          | Summary                                                                                                               |
+|--------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------|
+| version skew: ephemeralContainers case           | New pod field, need to be able to validate in same was containers and initContainers if field exists and is populated |
+| version skew: Migration from annotation to field | Need to be able to validate annotation (if present) or field (if it exists and is populated)                          |
+| CRD is deleted                                   | Nothing to type check against, but also means there are no coresponding custom resources                              |
+| CRD is in multiple clusters, but schema differs  | If policy author is aware of the schema variations, can they write policies that work for all the variations?         |
+| Validation of an aggregated API server type      | Main API server does not have type definitions                                                                        |
+
+Due to these complications, Type checking will be optional which can be activated by an expression-level
+`TypeChecking` field. Possible modes are:
```

Comment:
This is a lot of complexity to expose to the user. How are they supposed to anticipate whether or not their config will be subject to the problems listed above?

Reply:
I think upgrade and downgrade are things that need actual solutions.

Backwards-incompatible CRD changes we can treat as not our problem.

Therefore I think we can lock this to a particular setting and not permit the user to change it. "Informative" might be the right one.

Reply:
That's a fair point. Would you think allowing the user to disable type check can be helpful? For example, `has` operator can check if a field exists, take the `ephemeralContainers` example again.
`has(somePod.ephemeralContainers) && somePod.ephemeralContainers[0].image == "busybox"` would populate warnings in Kubernetes < 1.23, but because of the `has` check, the type check error is a false alarm.

Reply:
I think we need some plan for what to do around upgrades/downgrades. I'm not sure that's the best approach though...

Reply:
`Informative` mode is safe for upgrades/downgrades because it never fails the policy even if a type check error occurs. I will go ahead and make `Informative` the only mode to implement in this release. We can revisit this part in the next.

Reply:
I've updated the design to reflect the discussion. Please take another look when available. Thank you.